ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java

```diff
@@ -150,42 +162,69 @@ private void onBlockAdded(final BlockAddedEvent blockAddedEvent) {
                 if (r != null) {
                   LOG.info("Imported {} pending blocks", r.size());
                 }
+                if (t != null) {
+                  LOG.error("Error importing pending blocks", t);
+                }
               });
     } else {
 
-      LOG.trace("Not ready for import blocks found for {}", newBlock.getHeader().getNumber());
+      logAtTrace(
+          () ->
+              LOG.trace(
+                  "There are no pending blocks ready to import for block {}",
+                  newBlock.toLogString()));
+
+      maybeProcessNonAnnouncedBlocks(newBlock);
+    }
+
+    if (blockAddedEvent.getEventType().equals(EventType.HEAD_ADVANCED)) {
+      final long head = protocolContext.getBlockchain().getChainHeadBlockNumber();
+      final long cutoff = head + config.getBlockPropagationRange().lowerEndpoint();
+      pendingBlocksManager.purgeBlocksOlderThan(cutoff);
```

Comment:
~~I know this is unchanged from the original algorithm, but why would we purge future blocks?  the value of cutoff is going to be > than head right?  `purgeBlocksOlderThan(cutoff)` seems like it should be called on `head - lowerEndpoint` not `head + lowerEndpoint` ?~~

nvm: I just realized the lower bound is negative